pydata_google_auth/auth.py

```diff
@@ -311,6 +312,19 @@ def get_user_credentials(
     pydata_google_auth.exceptions.PyDataCredentialsError
         If unable to get valid user credentials.
     """
+
+    # If the user isn't explicitly asking to authenticate with a particular
+    # client ID, then we should try to authenticate the user with Colab-based
+    # credentials, if possible.
+    if client_id is None:
+        # default_project ignored for colab credentials
+        credentials, _ = get_colab_default_credentials(scopes)
+
+        if credentials and credentials.valid:
```

Comment:
why not simply remove this if clause and return? Then the line 388-399 kicks in and refreshes the credentials if needed?

Reply:
it would also require moving lines 328-386 under `if credentials is None` condition

Reply:
I really want to break early in this case. I don't want to risk trying to fetch other kinds of credentials.

Refresh is only really needed for `gcloud` and cached credentials because those credentials have a "refresh token" but not an "access token".

```suggestion
        # Break early to avoid trying to fetch any other kinds of credentials.
        # Prefer Colab credentials over any credentials based on the default
        # client ID.
        if credentials and credentials.valid:
```

Reply:
Oh, I probably don't need the `valid` piece for Colab credentials.

```suggestion
        if credentials:
```